File: apps/web/src/types/mloPreview.ts
export type MloPreview = {
  version: 1;
  name: string;
  archetypes: MloArchetype[];
  assets: MloAsset[];
  requiredDrawables: string[];
  requiredTextures: string[];
};

export type MloAsset = {
  name: string;
  format: "YDR" | "YFT";
  modelCount: number;
  boundsMin: [number, number, number];
  boundsMax: [number, number, number];
};

export type MloArchetype = {
  name: string;
  isMlo: boolean;
  drawableDictionary: string;
  textureDictionary: string;
  boundsMin: [number, number, number];
  boundsMax: [number, number, number];
  rooms: MloRoom[];
  portals: MloPortal[];
  entities: MloEntity[];
};

export type MloRoom = {
  index: number;
  name: string;
  boundsMin: [number, number, number];
  boundsMax: [number, number, number];
  floorId: number;
  portalCount: number;
  exteriorVisibilityDepth: number;
};

export type MloPortal = {
  index: number;
  roomFrom: number;
  roomTo: number;
  flags: number;
  opacity: number;
  corners: [number, number, number][];
  center: [number, number, number];
};

export type MloEntity = {
  index: number;
  name: string;
  archetype: string;
  position: [number, number, number];
  rotation: [number, number, number, number];
  scaleXY: number;
  scaleZ: number;
};

export type MloPortalPatch = {
  archetype: string;
  portalIndex: number;
  roomFrom: number;
  roomTo: number;
  flags: number;
  opacity: number;
  corners: [number, number, number][];
};

File: apps/engine/Contracts/MloPreview.cs
namespace FiveMesh.Engine.Contracts;

internal sealed record MloPreview(
    int Version,
    string Name,
    IReadOnlyList<MloArchetypePreview> Archetypes,
    IReadOnlyList<MloAssetPreview> Assets,
    IReadOnlyList<string> RequiredDrawables,
    IReadOnlyList<string> RequiredTextures
);

internal sealed record MloAssetPreview(
    string Name,
    string Format,
    int ModelCount,
    float[] BoundsMin,
    float[] BoundsMax
);

internal sealed record MloArchetypePreview(
    string Name,
    bool IsMlo,
    string DrawableDictionary,
    string TextureDictionary,
    float[] BoundsMin,
    float[] BoundsMax,
    IReadOnlyList<MloRoomPreview> Rooms,
    IReadOnlyList<MloPortalPreview> Portals,
    IReadOnlyList<MloEntityPreview> Entities
);

internal sealed record MloRoomPreview(
    int Index,
    string Name,
    float[] BoundsMin,
    float[] BoundsMax,
    int FloorId,
    uint PortalCount,
    int ExteriorVisibilityDepth
);

internal sealed record MloPortalPreview(
    int Index,
    uint RoomFrom,
    uint RoomTo,
    uint Flags,
    uint Opacity,
    float[][] Corners,
    float[] Center
);

internal sealed record MloEntityPreview(
    int Index,
    string Name,
    string Archetype,
    float[] Position,
    float[] Rotation,
    float ScaleXY,
    float ScaleZ
);
